### PR TITLE
libroach: test EncryptedEnv file ops.

### PR DESCRIPTION
**Pull request by mberhault**
We used to perform the operation on the file registry followed by the
underlying env.

This causes problems when using plaintext files as they don't have a
registry entry. This is fine for most operations, but LinkFile for
example fails if the new file exists. If we run before the Env method,
we have no way to tell without stating the file.

Add tests using a real env as the MemEnv doesn't really do any of those
correctly right. Also make use of experimental::filesystem (we're still
on C++11) for convenience.

### CODE REVIEW

File: c-deps/libroach/ccl/db_test.cc [hunk 1]
```diff
@@ -82,65 +82,3 @@ TEST(LibroachCCL, DBOpen) {
     DBClose(db);
   }
 }
-
-TEST(LibroachCCL, EncryptedEnv) {
```
Comment (mberhault) [inline c-deps/libroach/ccl/db_test.cc]: This is moved without modification (except for the name) to `encrypted_env_test.cc`.

File: c-deps/libroach/file_registry.cc [hunk 2]
```diff
@@ -168,6 +168,10 @@ rocksdb::Status FileRegistry::MaybeRenameEntry(const std::string& src, const std
   std::string newSrc = TransformPath(src);
   std::string newTarget = TransformPath(target);
 
+  if (newSrc == newTarget) {
```
Comment (mberhault) [inline c-deps/libroach/file_registry.cc]: Simple shortcut.

File: c-deps/libroach/file_registry.cc [hunk 3]
```diff
@@ -209,6 +213,7 @@ rocksdb::Status FileRegistry::MaybeLinkEntry(const std::string& src, const std::
   auto target_key = registry_->files().find(newTarget);
   if ((src_key == registry_->files().cend()) && (target_key == registry_->files().cend())) {
     // Entries do not exist: do nothing.
+    return rocksdb::Status::OK();
```
Comment (mberhault) [inline c-deps/libroach/file_registry.cc]: Forgot this one. It ends up just doing a needless write.

File: c-deps/libroach/testutils.h [hunk 4]
```diff
@@ -67,8 +67,32 @@ class FakeTimeEnv : public rocksdb::EnvWrapper {
...
+// 'err_msg' (regexp full match).
+#define ASSERT_ERR(status, err_msg)\
+  {\
+    auto s(testutils::compareErrorMessage(status, err_msg, false)); \
+    ASSERT_TRUE(s.ok()) << s.getState();\
+  }
+
+// If err_msg is empty, status must be ok. Otherwise, the status message must match
+// 'err_msg' (regexp partial match).
+#define EXPECT_PARTIAL_ERR(status, err_msg)\
```
Comment (mberhault) [inline c-deps/libroach/testutils.h]: Added partial error matches. This is needed if part of the error is binary (eg: when I dump file contents) as `.` in `std::regex` does not match newlines.

**Comment by bdarnell** [comment]
:lgtm:

---

Review status: 1 of 9 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[c-deps/libroach/CMakeLists.txt, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26447#-LEKjWPs2nxmqJfo_x8b:-LEKjWPtB7nin3-lQSp9:bgtnqdz) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc49b87c6defcd2c827092c1e56f6196dfa5d997/c-deps/libroach/CMakeLists.txt#L131)):*
> ```CMake
>     target_link_libraries(${tname}
>       roachccl
>       stdc++fs
> ```

Is this guaranteed to be present on every compiler we support?

---

*[c-deps/libroach/ccl/testutils.cc, line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26447#-LEKlgG-ES6uMrnLOnQd:-LEKlgG-ES6uMrnLOnQe:b-57r9tf) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc49b87c6defcd2c827092c1e56f6196dfa5d997/c-deps/libroach/ccl/testutils.cc#L68)):*
> ```cpp
>   char* tmpl = new char[strlen(fs_dir.c_str()) + 1];
>   tmpl[strlen(fs_dir.c_str())] = '\0'; /* stupid null-terminated string */
>   strcpy(tmpl, fs_dir.c_str());
> ```

Doesn't this need to be `strncpy` because `fs_dir` is not null-terminated?

It's a little silly that C doesn't have a standard accessor for the temp directory and C++ doesn't have a version of `mkdtemp`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26447#-:-LEKmRL58ilenvllh2hd:bnfp4nl)*

**Comment by mberhault** [comment]
I've moved `TempDirHandler` to the non-ccl `testutils.h`.

---

Review status: 1 of 10 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[c-deps/libroach/CMakeLists.txt, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26447#-LEKjWPs2nxmqJfo_x8b:-LEKoEPmCJlE4xgxXl43:b-sseakw) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc49b87c6defcd2c827092c1e56f6196dfa5d997/c-deps/libroach/CMakeLists.txt#L131)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is this guaranteed to be present on every compiler we support?
</blockquote></details>

the ones I've tried so far. C++11 compilers "usually" include the experimental filesystem stuff. If it doesn't, we only break people trying to run the libroach tests. 

---

*[c-deps/libroach/ccl/testutils.cc, line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26447#-LEKlgG-ES6uMrnLOnQd:-LEKoQ2p9pLhnoMbKz4c:by0fn61) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc49b87c6defcd2c827092c1e56f6196dfa5d997/c-deps/libroach/ccl/testutils.cc#L68)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Doesn't this need to be `strncpy` because `fs_dir` is not null-terminated?

It's a little silly that C doesn't have a standard accessor for the temp directory and C++ doesn't have a version of `mkdtemp`.
</blockquote></details>

Done. It has the advantage of doing the null termination for me.
And yes, it's rather annoying. It would be even worse without `experimental::filesystem`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26447#-:-LEKoWXi530efO7Gnewe:b-jsx5uv)*

**Pull request merged**